Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/distance.h

```diff
@@ -116,67 +129,580 @@ sample_triangles(const TriangleRange& triangles, double distance, OutputIterator
+ * @tparam OutputIterator a model of `OutputIterator` 
+ *  holding objects of the same point type as
+ *  the value type of the internal vertex point map of `tm`
+ * @tparam Sampling_method defines the sampling method. Possible values are:
+     - `RANDOM_UNIFORM`: points are generated in a random and uniform way, depending on the area of each triangle.
+     - `GRID`: points are generated on a grid in each triangle, with a minimum of one point per triangle.
+     - `MONTE_CARLO`: points are generated randomly in each triangle. The number of points per triangle is proportional to the triangle area with a minimum of 1.
+ *
+ * @param tm the triangle mesh that will be sampled
+ * @param parameter depends on `method`:
+     - `RANDOM_UNIFORM` and `MONTE_CARLO`: the number of points per squared area unit
+     - `GRID`: the distance between two consecutive points in the grid
+ *
+ * @param out output iterator to be filled with sampled points
+ * @param np an optional sequence of \ref namedparameters among the ones listed below
+ * @param method defines the sampling method
+ *
+ * \cgalNamedParamsBegin
+ *    \cgalParamBegin{vertex_point_map}
+ *    the property map with the points associated to the vertices of `tm`. If this parameter is omitted,
+ *    an internal property map for `CGAL::vertex_point_t` should be available for `TriangleMesh` \cgalParamEnd
+ *    \cgalParamBegin{geom_traits} an instance of a geometric traits class, model of `PMPDistanceTraits`\cgalParamEnd
+ * \cgalNamedParamsEnd
+ */
+template<class OutputIterator, class TriangleMesh, class NamedParameters>
+OutputIterator
+sample_triangle_mesh(const TriangleMesh& tm,
+                           double parameter,
+                           OutputIterator out,
+                           NamedParameters np,
+                           Sampling_method method = RANDOM_UNIFORM)
 {
-  std::vector<typename Kernel::Point_3> sample_points;
-  sample_triangles<Kernel>( triangles_1,
-                            targeted_precision,
-                            std::back_inserter(sample_points) );
-  // std::ofstream out("/tmp/samples.xyz");
-  // std::copy(sample_points.begin(), sample_points.end(), std::ostream_iterator<typename Kernel::Point_3>(out,"\n"));
+    typedef typename GetGeomTraits<TriangleMesh,
+            NamedParameters>::type Geom_traits;
+
+    typedef typename GetVertexPointMap<TriangleMesh,
+            NamedParameters>::const_type Vpm;
 
-  typedef typename TriangleRange2::const_iterator Triangle_iterator;
-  typedef AABB_triangle_primitive<Kernel, Triangle_iterator> Primitive;
+    Vpm pmap = choose_param(get_param(np, vertex_point),
+                           get_const_property_map(vertex_point, tm));
+    typedef Creator_uniform_3<typename Geom_traits::FT,
+                              typename Geom_traits::Point_3> Creator;
+  switch(method)
+  {
+    case RANDOM_UNIFORM:
+    {
+      std::size_t nb_points = std::ceil( to_double(
```

Comment:
@sloriot : Is there only a `static_cast` missing to  avoid a VC++ warning, or isn't there even a problem when the area is very small?